### PR TITLE
Additional client information websocket

### PR DESCRIPTION
**Pull request by achim-k**
Latches a list of all connected clients, where each client is described by
- the client's IP address
- the time the client connected

This only works for the websocket server for now. Let me know if this is something useful. If so, then I will look into adding this functionality also to the other backends.

### CODE REVIEW

**Comment by jihoonl** [comment]
Looks good. Thanks.
I think `open` and `on_close` are not thread safe. Could you put a lock on `add_client`, and `remove_client`?

File: rosbridge_server/src/rosbridge_server/client_mananger.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+          client = ConnectedClient()
+          client.ip_address = ip_address
+          client.connection_time = rospy.Time.now()
+          self._clients[client_id] = client
+          self.__publish()
+
+    def remove_client(self, client_id, ip_address):
+        with self._lock:
+          self._clients.pop(client_id, None)
+          self.__publish()
```
Comment (jihoonl) [inline rosbridge_server/src/rosbridge_server/client_mananger.py]: Small NITS: wrong indentation.

File: rosbridge_server/src/rosbridge_server/client_mananger.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    def __publish(self):
+        msg = ConnectedClients()
+        msg.clients = list(self._clients.values())
+        self._conn_clients_pub.publish(msg.clients)
+        self._client_count_pub.publish(len(msg.clients))
+
+    def add_client(self, client_id, ip_address):
+        with self._lock:
+          client = ConnectedClient()
```
Comment (jihoonl) [inline rosbridge_server/src/rosbridge_server/client_mananger.py]: here too

File: rosbridge_server/src/rosbridge_server/client_mananger.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+from rosbridge_msgs.msg import ConnectedClient, ConnectedClients
+from std_msgs.msg import Int32
+
+
+class ClientManager:
+    def __init__(self):
+        # Publisher for number of connected clients
+        self._client_count_pub = rospy.Publisher(
+          'client_count', Int32, queue_size=10, latch=True)
+        self._client_count_pub.publish(0)
```
[the remark this answers is not in the document]
Comment (achim-k) [reply rosbridge_server/src/rosbridge_server/client_mananger.py]: good catch!

File: rosbridge_server/src/rosbridge_server/client_mananger.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        self._conn_clients_pub = rospy.Publisher(
+          'connected_clients', ConnectedClients, queue_size=10, latch=True)
+        self._client_count = 0
+        self._clients = {}
+        self.__publish()
+
+    def __publish(self):
+        msg = ConnectedClients()
+        msg.clients = list(self._clients.values())
+        self._conn_clients_pub.publish(msg.clients)
```
[the remark this answers is not in the document]
Comment (achim-k) [reply rosbridge_server/src/rosbridge_server/client_mananger.py]: this should indeed just publish `msg` :+1:

File: rosbridge_msgs/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,21 @@
+cmake_minimum_required(VERSION 2.8.3)
+project(rosbridge_msgs)
+
+find_package(catkin REQUIRED COMPONENTS
+  genmsg std_msgs
```
[the remark this answers is not in the document]
Comment (achim-k) [reply rosbridge_msgs/CMakeLists.txt]: Good remark. Fixed it now according to http://docs.ros.org/kinetic/api/catkin/html/howto/format2/building_msgs.html

**Comment by achim-k** [comment]
This PR has quite some commits now. I can rebase and squash my commits if that desired.

**Comment by jihoonl** [comment]
Note that this PR introduced a new dependency on rosbridge_server. So, we need to release with minor patch bump for the next release.

**Comment by ramezanifar** [comment]
I would like to prevent certain IP addresses from connecting to rosbridge websocket server.
How can I do that? Is there a way to pass a list of allowed IP addresses when I launch the server?
Thank you

**Comment by mvollrath** [comment]
@ramezanifar system firewall features are good for that, if you are using Ubuntu look at `ufw`.

**Pull request merged**